src/main/java/seedu/address/model/tut/Tutorial.java

```diff
@@ -11,182 +11,402 @@
-    }
+    /**
+     * Returns the tutorial class information.
+     *
+     * @return The tutorial class.
+     * @throws NoTutorialException if called on a None instance.
+     */
+    public abstract TutorialClass getTutorialClass();
 
-    public boolean isValidTutorialDate(TutDate tutorialDate) {
-        return tutorialDate.isValid();
+    /**
+     * Marks attendance for a specific student on a given tutorial date.
+     *
+     * @param student The student to mark attendance for.
+     * @param tutDate The tutorial date.
+     * @throws TutDateNotFoundException if the tutorial date is not found.
+     * @throws NoTutorialException if called on a None instance.
+     */
+    public abstract void markAttendance(Student student, TutDate tutDate);
+
+    /**
+     * Checks if a student is in the list of students.
+     *
+     * @param student The student to check.
+     * @return True if the student is in the list; otherwise false.
+     */
+    public abstract boolean studentInList(Student student);
+
+    /**
+     * Deletes a student from the tutorial.
+     *
+     * @param student The student to be removed.
+     */
+    public abstract void deleteStudent(Student student);
+
+    private static Tutorial exist(TutName tutName, TutorialClass tutorialClass) {
+        return new Exist(tutName, tutorialClass);
     }
 
-    public List<TutDate> getTutDates() {
-        return new ArrayList<>(tutDates.values());
+    /**
+     * Creates a tutorial that represents no tutorial.
+     *
+     * @return A None tutorial object.
+     */
+    public static Tutorial none() {
+        return None.none();
     }
 
-    public TutorialClass getTutorialClass() {
-        return tutorialClass;
+    /**
+     * Creates an instance of an existing tutorial.
+     *
+     * @param tutName The name of the tutorial.
+     * @param tutorialClass The class of the tutorial.
+     * @return A new tutorial object.
+     */
+    public static Tutorial of(TutName tutName, TutorialClass tutorialClass) {
+        requireNonNull(tutName);
+        requireNonNull(tutorialClass);
+        return Tutorial.exist(tutName, tutorialClass);
     }
 
     /**
-     * Marks the attendance for the student for the particular tutorial date.
+     * Represents a None type of tutorial (non-existent).
      */
-    public void markAttendance(Student student, TutDate tutorialDate) throws TutDateNotFoundException {
-        if (!isValidTutorialDate(tutorialDate)) {
-            throw new TutDateNotFoundException();
+    private static final class None extends Tutorial {
+        private static final None none = new None();
+
+        private final List<Student> students = new ArrayList<>();
+
+        public static None none() {
+            return none;
         }
-        if (!studentInList(student)) {
-            add(student);
+
+        @Override
+        public void add(Student student) {
+            if (student != null && !students.contains(student)) {
```

Comment:
Maybe using requireNonNull here for student is a better option? If not the student will just not get added and the null error is not thrown.